### PR TITLE
Pretraining GATr

### PR DESCRIPTION
**Pull request by spinjo**
Changes
- Renamed toptagging -> tagging; restructured config files for tagging
- Rewrote tagging embedding code
- JetClass experiment: Load data with miniweaver (adapted from https://github.com/hqucms/weaver-core), new experiment class
- TopTaggingFineTune experiment: Load JetClass-pretrained model and finetune on toptagging
- Removed absolute paths from config files

TODO
- QuarkGluonFineTune experiment

### CODE REVIEW

File: experiments/baselines/ds_mlp.py
Comment (spinjo) [inline experiments/baselines/ds_mlp.py:1]: Where is this coming from?

File: experiments/base_experiment.py [hunk 1]
```diff
@@ -459,13 +459,15 @@ def train(self):
         patience = 0
 
         # main train loop
+        """
```
Comment (spinjo) [inline experiments/base_experiment.py:462]: Why remove this logger output?

File: config/ttbar.yaml [hunk 2]
```diff
@@ -64,4 +64,4 @@ defaults:
 - classifier: classifier
 - default
 - hydra
-- _self_
```
Comment (spinjo) [inline config/ttbar.yaml:67]: You should get rid off this kind of changes, eg with 'git checkout main file' to just pull this file from the main branch. Same for the other config files.

File: data/utils/lr_finder.py
Comment (spinjo) [inline data/utils/lr_finder.py:1]: I'd guess we don't use weaver's lr_finder. Would be good if you check which parts of weaver we're actually using, and only keep those - just to keep everything readable.

File: experiments/base_experiment.py
Comment (spinjo) [inline experiments/base_experiment.py:1]: This change will mess up the other experiments. Could you adapt the JetClass experiment accordingly?

File: experiments/toptagging/experiment.py [hunk 3]
```diff
@@ -285,8 +377,15 @@ def _validate(self, step):
         return metrics["bce"]
 
     def _batch_loss(self, batch):
-        y_pred = self.model(batch)
-        loss = self.loss(y_pred, batch.label.to(self.dtype))
+        LOGGER.info(f"The batch is {batch}")
+        if self.cfg.exp_type == "jctagging":
```
Comment (spinjo) [inline experiments/toptagging/experiment.py:381]: Your if-statement approach makes it very hard to keep an overview of our different kinds of experiments. Instead of adding if statements to the abstract TaggingExperiment class, you should have this functionality in the JetClassExperiment defined below, specifically in this case you can just overwrite the _batch_loss function of that class.

File: experiments/toptagging/experiment.py [hunk 4]
```diff
@@ -270,6 +359,9 @@ def plot(self):
     def _init_loss(self):
         self.loss = torch.nn.BCEWithLogitsLoss()
 
+    def _init_loss_pretrain(self):
+        self.loss = torch.nn.CrossEntropyLoss()
```
Comment (spinjo) [inline experiments/toptagging/experiment.py:363]: This should go into the JetClassExperiment class

File: experiments/toptagging/experiment.py [hunk 5]
```diff
@@ -134,6 +190,39 @@ def _init_dataloader(self):
             f"batch_size={self.cfg.training.batchsize} (training), {self.cfg.evaluation.batchsize} (evaluation)"
         )
 
+    def _init_dataloader_pretrain(self):
```
Comment (spinjo) [inline experiments/toptagging/experiment.py:193]: This should go into the JetClassExperiment class

File: experiments/toptagging/experiment.py [hunk 6]
```diff
@@ -111,18 +115,70 @@ def _init_data(self, Dataset, data_path):
         dt = time.time() - t0
         LOGGER.info(f"Finished creating datasets after {dt:.2f} s = {dt/60:.2f} min")
 
+    def _init_data_pretrain(self, Dataset):
```
Comment (spinjo) [inline experiments/toptagging/experiment.py:118]: This should go into the JetClassExperiment class

File: experiments/toptagging/experiment.py
Comment (spinjo) [inline experiments/toptagging/experiment.py:1]: To keep things clear and readable also for our old experiments, I suggest you do not touch this file at all, and instead move everything to a new file `experiments/toptagging/jetclassexperiment.py` where you have the derived class at the bottom of this file and all the specialised functionality related to it. I would do the same for the TopTaggingExperiment and QGTaggingExperiment, but these classes are so short that its not really worth it.

File: experiments/toptagging/wrappers.py [hunk 7]
```diff
@@ -2,13 +2,14 @@
...
-from gatr.interface import extract_scalar
+from gatr.interface import extract_scalar, embed_vector
+from experiments.toptagging.dataset import embed_beam_reference
 from xformers.ops.fmha import BlockDiagonalMask
 
 
 def attention_mask(batch, force_xformers=True):
     """
-    Construct attention mask that makes surfe that objects only attend to each other
+    Construct attention mask that makes sure that objects only attend to each other
```
Comment (spinjo) [inline experiments/toptagging/wrappers.py:12]: Thanks! :)

File: experiments/toptagging/wrappers.py [hunk 8]
```diff
@@ -31,6 +32,199 @@ def attention_mask(batch, force_xformers=True):
...
+        self.mean_aggregation = mean_aggregation
+        self.add_pt = add_pt
+        self.force_xformers = force_xformers
+        self.add_jet_momentum = add_jet_momentum
+        self.beam_reference = beam_reference
+        self.add_time_reference = add_time_reference
+        self.two_beams = two_beams
+        self.beam_token = beam_token
+
+    def forward(self, batch):
```
Comment (spinjo) [inline experiments/toptagging/wrappers.py:111]: Instead of having this big method with the wild (and probably very expensive) for loop, we could do this in the dataset class, similar to the TopTaggingDataset and QGTaggingDataset. (As a side note, we could unify these two a bit because they share a lot of code. But low prio.) 

Also, it looks to me like nothing JetClass-specific is happening here anymore, i.e. the data is already in the same format as for our other tagging experiments. The only thing you do is to bring everything into the torch_geometric format. Can you confirm?

**Review by spinjo — COMMENTED** [review]
Thanks for all the work! 

I added many comments, noting everything that came to mind. The main points are:
- There is no reason to do changes that mess up other experiments, e.g. your changes to some config files, run.py and base_experiment.py. I suspect that you didn't start your changes off the main branch, but from a local code version. 
- I didn't study the weaver-related changes. It would be good if you go through the files you import from weaver, and remove everything that we're not using (at least to first approximation). This makes it easier for others to follow what is going on, right now this is a bit overwhelming.
- Instead of modifying the abstract JetTaggingExperiment class, you should modify the JetClassTaggingExperiment to keep things modular. You can also create a new file jetclassexperiment.py or so for that, that would be even better.
- Instead of creating a default torch dataloader and bringing everything into torch_geometric format within the wrapper, you should directly create a torch_geometric dataloader as we do for the other experiments. This is easier to read, and probably also faster (the torch_geometric dataloader does not use a for loop to construct the batch).

File: requirements.txt
[the remark this answers is not in the document]
Comment (vbpla) [reply requirements.txt:1]: I have removed all redundancies. For unused packages, I checked the ones that are explicitly imported in the scripts I am using and removed the ones I didn't see. The ones that remain are:

- tqdm>=4.41.0
- PyYAML>=5.4
- awkward0>=0.15.5
- uproot>=4.2.0,<5.2.0
- awkward>=1.8.0

File: experiments/toptagging/wrappers.py [hunk 9]
```diff
@@ -31,6 +32,199 @@ def attention_mask(batch, force_xformers=True):
         mask = mask.materialize(shape=(len(batch.batch), len(batch.batch)))
     return mask
 
+def attention_mask_pretrain(attention_mask_list, force_xformers=True):
```
[the remark this answers is not in the document]
Comment (vbpla) [reply experiments/toptagging/wrappers.py:38]: It's the argument of the function. The original function calls batch.batch within the function and with my setup I compute that information by hand and give it directly as an argument. If the original function was changed to take batch.batch as an argument, the functions could be merged.

File: experiments/toptagging/wrappers.py [hunk 10]
```diff
@@ -31,6 +32,199 @@ def attention_mask(batch, force_xformers=True):
...
+    """
+    bincounts = torch.bincount(attention_mask_list).tolist()
+    mask = BlockDiagonalMask.from_seqlens(bincounts)
+    if not force_xformers:
+        # materialize mask to torch.tensor (only for testing purposes)
+        mask = mask.materialize(shape=(len(attention_mask_list), len(attention_mask_list)))
+    return mask
+
+
+class TopTaggingTransformerWrapper(nn.Module):
```
[the remark this answers is not in the document]
Comment (vbpla) [reply experiments/toptagging/wrappers.py]: I removed it already in the latest commit.

File: experiments/baselines/ds_mlp.py
Comment (vbpla) [reply experiments/baselines/ds_mlp.py:1]: Removed already.

File: experiments/base_experiment.py (hunk 1 above)
Comment (vbpla) [reply experiments/base_experiment.py:454]: You cannot call len(self.train_loader) if you build the dataloaders with the SimpleIterDataset class

File: config/ttbar.yaml (hunk 2 above)
Comment (vbpla) [reply config/ttbar.yaml:67]: Solved.

File: data/utils/lr_finder.py
Comment (vbpla) [reply data/utils/lr_finder.py:1]: You are probably right. For me this is low priority. Once everything is running, I will trim down the weaver folder accordingly.

File: experiments/toptagging/wrappers.py (hunk 8 above)
Comment (vbpla) [reply experiments/toptagging/wrappers.py:91]: Yes, that is essentially what I do. And I also could somehow move all of this to the dataset file. However, I cannot get rid of the for loop by developing a JCTaggingDataset class akin to the preeexistent ones, because I would need to load the whole dataset in memory. SimpleIterDataset by default loads a full batch at once, so we will have to adapt it so that the global token, beam, etc., are added without the for loop. It should be definitely possible, but it was non-trivial the first time I tried doing it.

File: experiments/toptagging/experiment.py
Comment (vbpla) [reply experiments/toptagging/experiment.py:1]: Agreed, I will do this right away.

File: experiments/base_experiment.py (hunk 1 above)
Comment (spinjo) [reply experiments/base_experiment.py:454]: Fair point. It would still be good to have all this information. I think there is a way to implement the len(self.train_loader) thing, namely you have to implement the __len__ method for the SimpleIterDataset properly. A quick fix would be to do something like num_batches = 1239842/batchsize if exp_type=jctagging else len(self.train_loader) where you have to hard-code the number of events in the jetclass dataset.

File: experiments/base_experiment.py [hunk 11]
```diff
@@ -70,14 +70,14 @@ def full_run(self):
...
         # save config
         LOGGER.debug(OmegaConf.to_yaml(self.cfg))
-        self._save_config("config.yaml", to_mlflow=True)
+        self._save_config("config.yaml", to_mlflow=False)
         self._save_config(f"config_{self.cfg.run_idx}.yaml")
 
         self.init_physics()
         self.init_model()
         self.init_data()
-        self._init_dataloader()
```
Comment (spinjo) [inline experiments/base_experiment.py:79]: Why do you have to change these names?

File: experiments/base_experiment.py [hunk 12]
```diff
@@ -510,13 +512,13 @@ def cycle(iterable):
                 )
 
         dt = time.time() - self.training_start_time
-        LOGGER.info(
```
Comment (spinjo) [inline experiments/base_experiment.py:513]: All this stuff can be done dynamically if we somehow extract len(self.train_loader), see comment above.

File: experiments/toptagging/experiment.py [hunk 13]
```diff
@@ -257,7 +258,7 @@ def plot(self):
             plot_dict["val_metrics"] = self.val_metrics
         plot_mixer(self.cfg, plot_path, title, plot_dict)
 
-    def _init_loss(self):
```
Comment (spinjo) [inline experiments/toptagging/experiment.py:260]: Again, I would not change this name.

File: experiments/toptagging/experiment.py [hunk 14]
```diff
@@ -271,6 +272,7 @@ def _validate(self, step):
             metrics = self._evaluate_single(
                 self.val_loader, "val", mode="val", step=step
             )
+
```
Comment (spinjo) [inline experiments/toptagging/experiment.py:275]: You could remove these extra white lines with 'git checkout -p main file.py', that makes it more readable.

File: gatr/layers/linear.py [hunk 15]
```diff
@@ -9,6 +9,8 @@
 from gatr.interface import embed_scalar
 from gatr.primitives.linear import NUM_PIN_LINEAR_BASIS_ELEMENTS, equi_linear
 
+from experiments.logger import LOGGER
```
Comment (spinjo) [inline gatr/layers/linear.py:12]: I think there is no need for doing this. You can just reset this to the main branch (and same for gatr.py below)

File: run.py [hunk 16]
```diff
@@ -7,16 +7,18 @@
     z5gExperiment,
 )
 from experiments.toptagging.experiment import QGTaggingExperiment
+from experiments.toptagging.jetclassexperiment import JetClassTaggingExperiment
 
-
-@hydra.main(config_path="config", config_name="amplitudes", version_base=None)
+@hydra.main(config_path="config", config_name="jctagging", version_base=None)
```
Comment (spinjo) [inline run.py:12]: Before merging into main, I would undo this change again to have 'amplitudes' as the default. Happy to keep it as it is until then. Note that you can specify the config_name with --config_name=XYZ or -cn XYZ

**Review by spinjo — COMMENTED** [review]
Good work! Mentioned some minor things that are quick to fix. I'll study the JetClass file more carefully later on.

File: experiments/tagging/embedding.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+        )
+        is_global[global_idxs] = True
+        multivectors_buffer = multivectors.clone()
+        multivectors = torch.zeros(
+            is_global.shape[0],
+            *multivectors.shape[1:],
+            dtype=multivectors.dtype,
+            device=multivectors.device,
+        )
+        multivectors[~is_global] = multivectors_buffer
```
Comment (vbpla) [inline experiments/tagging/embedding.py:118]: Before this line, we should add the following lines to include a 1 in the scalar component of the multivector global token:

```suggestion
        global_mv = torch.zeros((1, multivectors.shape[-1]), device=multivectors.device)
        global_mv[:,0] = 1
        multivectors[is_global] = global_mv.repeat(batchsize * num_global_tokens, 1).unsqueeze(-2)
        multivectors[~is_global] = multivectors_buffer

File: experiments/tagging/miniweaver/kitchensink.yaml [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+         - [part_deta, null]
+         - [part_dphi, null]
+   pf_vectors:
+      length: 128
+      pad_mode: constant
+      vars: 
+         - [part_px, null]
+         - [part_py, null]
+         - [part_pz, null]
+         - [part_energy, null]
```
Comment (vbpla) [inline experiments/tagging/miniweaver/kitchensink.yaml:53]: This should be reordered.

```suggestion
         - [part_energy, null]
         - [part_px, null]
         - [part_py, null]
         - [part_pz, null]
```

File: experiments/tagging/miniweaver/default.yaml [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+         - [part_dz, null]
+         - [part_dzerr, 0, 1, 0, 1]
+   pf_vectors:
+      length: 128
+      pad_mode: constant
+      vars: 
+         - [part_px, null]
+         - [part_py, null]
+         - [part_pz, null]
+         - [part_energy, null]
```
Comment (vbpla) [inline experiments/tagging/miniweaver/default.yaml:40]: This should be reordered.

```suggestion
         - [part_energy, null]
         - [part_px, null]
         - [part_py, null]
         - [part_pz, null]
```

File: experiments/tagging/miniweaver/pid.yaml [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+         - [part_isElectron, null]
+         - [part_isMuon, null]
+   pf_vectors:
+      length: 128
+      pad_mode: constant
+      vars: 
+         - [part_px, null]
+         - [part_py, null]
+         - [part_pz, null]
+         - [part_energy, null]
```
Comment (vbpla) [inline experiments/tagging/miniweaver/pid.yaml:30]: This should be reordered.

```suggestion
         - [part_energy, null]
         - [part_px, null]
         - [part_py, null]
         - [part_pz, null]
```

File: experiments/tagging/jetclassexperiment.py [hunk 21]
```diff
@@ -0,0 +1,275 @@
...
+            "pin_memory": True,
+            "num_workers": self.cfg.jc_params.num_workers,
+            "persistent_workers": self.cfg.jc_params.num_workers > 0
+            and self.cfg.jc_params.steps_per_epoch is not None,
+        }
+        self.train_loader = DataLoader(
+            dataset=self.data_train,
+            batch_size=self.cfg.training.batchsize,
+            drop_last=True,
+            **self.loader_kwargs,
```
Comment (vbpla) [inline experiments/tagging/jetclassexperiment.py:138]: Maybe add "shuffle=True" here.

Comment (spinjo) [reply experiments/tagging/jetclassexperiment.py:138]: Tried it, apparently it does not make sense with `IterableDataset` objects which in turn makes sense (you randomly select data from files)
`ValueError: DataLoader with IterableDataset: expected unspecified shuffle option, but got shuffle=True`

File: experiments/tagging/embedding.py (hunk 17 above)
Comment (spinjo) [reply experiments/tagging/embedding.py:118]: This information is already there in `scalars` (see the `token_idx` a few lines below), so no need to include it a second time. 

Your suggestion only works if we have 1 global token (`data.score_token=false`), in case `data.score_token=true`/10 global tokens you would need 10 multivector channels to have enough scalar channels where you can embed this information (similar to how we do it with the `token_idx` right now in `scalar` channels).
